### ISSUE TITLE
Nitro Tasks API

### ISSUE THREAD

**Issue by pi0** [nitrojs/nitro#1974]
Nitro tasks allow on-off operations in runtime.

Docs: https://nitro.unjs.io/guide/tasks

- [x] Initial Implementation (#1929)
- [x] Write documentation section (PR welcome!)
- [x] Stabilize API / return values
- [x] Integrate with Cron Tasks / Background Jobs
- [ ] https://github.com/unjs/nitro/issues/2633
- [x] Vercel support
- [ ] Ability to run tasks using CLI against production build
- [ ] Task hooks (on dev start, production start and production build)
- [ ] Ability to schedule one-time run of a task for the future
- [ ] Allow optionally allow running multiple instances of task

**Comment by imcm7** [comment]
But I thought it was about cron task. Mikro-orm has own commands like db:migrate  and other.

**Comment by pi0** [comment]
@imcm7 check out task list it is on the roadmap :)

**Comment by imcm7** [comment]
@pi0 I have a big trouble, I use mikro orm and form it with orm.em.fork(), but it have a [middleware for that](https://github.com/mikro-orm/mikro-orm/blob/cccc7561f9dc3c90f22ad293fceb7b58a67dabea/packages/core/src/utils/RequestContext.ts#L2)  but it nit work.

I do something like

```
return fromNodeMiddleware(async (
            req: IncomingMessage,
            res: ServerResponse<IncomingMessage>,
            next: (...args: any[]) => void
) =>  {
           RequestContext.create(em, next);
}
        
```

And I get result with in middleware

```
  private static storage = new AsyncLocalStorage<RequestContext>();
  console.log('>>>>',  this.storage.getStore());
  
```

But no result in mikro-rom when it try to load forked entity manager.


useEvent() say no nitro instance (may be because is called from plugin) and for that is need to change mikro-orm code.


And with that don't know what to do

```
  export const nitroAsyncContext = getContext<NitroAsyncContext>("nitro-app", {
    asyncContext: import.meta._asyncContext,
    AsyncLocalStorage: import.meta._asyncContext ? AsyncLocalStorage : undefined,
  });


  return nitroAsyncContext.use();
    
```


How to combine 2 AsyncLocalStorage

**Comment by stafyniaksacha** [comment]
Do you plan to add a build time task runner? (e.g: `node .output/server/tasks.mjs run db:migrate --payload "{}"`)

**Comment by Aareksio** [comment]
Tasks should probably be async, currently the `run` method expects synchronous code and [is not awaited](https://github.com/unjs/nitro/blob/c8e85490c2895f568bdca22a52ff94befff78b45/src/runtime/task.ts#L58).

**Comment by manniL** [comment]
@Aareksio check https://github.com/unjs/nitro/issues/2021 ☺️

**Comment by Rigo-m** [comment]
If I use a nitro task in, let's say, Netlify will it leverage netlify background functions?

**Comment by cosbgn** [comment]
I'm also interested on this. I would like to have for example my normal code on vercel-edge but run tasks on vercel lambda (5min timeout vs 25 seconds)

**Comment by imcm7** [comment]
How to disable it, because i use dir tasks for other stuff?

**Comment by brendonmatos** [comment]
It is planned to support running these tasks from cli without running the server?

**Comment by pi0** [comment]
Nitro already has a **CLI** preset too! However it is tricky to tell now when we can do support it this way now.

**Comment by robin-dongbin** [comment]
Any plan for support seconds scheduling?

**Comment by robin-dongbin** [comment]
> @robin-dongbin You mean(意味着) this [nitro.unjs.io/guide/tasks#scheduled-tasks](https://nitro.unjs.io/guide/tasks#scheduled-tasks) or something different? (like a pattern(模式) for <1 minute(分钟) schedule?)

Yes, that's what I want. Is it possible? I know that cron do not support that, but Laravel has support this last year.

**Comment by go4cas** [comment]
@pi0 ... I have just tested the experimental scheduled task feature. It works really well! Thanks!

Just one question: would it be possible to have a config setting to execute a scheduled task immediately after the cron is scheduled? Example: a task running every 10 minutes: executes once the server starts, and then every 10 minutes after that? I saw in the `croner` docs that there is a `trigger` method, but not sure how it would fit into the nitro tasks?

**Comment by pi0** [comment]
Maybe use a Nitro plugin to triger them on startup? (also `init:` / `startup:` tasks are supported to have similar effect in the future). Also you can execute a task from another.

**Comment by JCtapuk** [comment]
> @go4cas Maybe use a Nitro plugin to triger them on startup? (also `init:` / `startup:` tasks are planned to be supported to have similar effect in the future). Also you can execute a task from another.
> 
> @robin-dongbin I also wonder if you need to trigger something this frequently, maybe a `setInterval` inside nitro plugins to run task with `runTask` might be simpler method with giving full control on executation?


But won’t there be conflicts if a node in a cluster with 8 processes and all 8 processes will cause a task, but this behavior should not happen? It turns out you need to find someone who has a free worker and give him a task from the master?

**Comment by pi0** [comment]
For time being, scheduled tasks are not supported in cluster mode the exactly meet this requirement (share a **safe lock** among them) and it is doable.

Regardless inside your interval you can call `runTask()` in single node, it currently guarantees only it is running once and when we support cluster, it would be safe too.

**Comment by brendonmatos** [comment]
I apologize if this isn't the correct forum. Could you please clarify how to execute tasks within a Nuxt project from the command line? Should I use the Nuxt CLI to trigger tasks indirectly via URLs, or we are supposed to use the Nitro CLI directly?

I apologize if this isn't the correct forum. But how is supposed to run those tasks in Nuxt project from CLI? Nuxt CLI will be implemented to invoke those tasks from the URL or we should run directly from nitro cli?

**Comment by pi0** [comment]
Yes.

**Comment by passionate-bram** [comment]
With async tasks, and enforcing task payloads to be serializable early on it will be possible in the future to dispatch tasks to queues. Shamelessly taking the concept from Laravel: https://laravel.com/docs/10.x/queues

The nice thing about that way of working is that when queues are configurable (e.g. redis, some AWS service, etc.) then the dispatch mechanic becomes really usefull for dispatching work to different parts of your architecture.

**Comment by pi0** [comment]
`payload` should be serializable indeed in preparation for future ideas like queues. and it is by design separated from `context` that is non seriazable for other stuff. PR welcome to clarify in the docs.

**Comment by GerryWilko** [comment]
@pi0 This tasks API looks fantastic thank you so much for adding this to Nitro.

Anyone who is trying to get early access to this feature and using a workspace with multiple nuxt projects may run into issues trying to get nuxt to pickup the correct nitro-imports. In my case I managed to force it to work with an overrides in the project `package.json`:

```json
  "pnpm": {
    "overrides": {
      "nitropack": "npm:nitropack-nightly@latest"
    }
  }
```

Thought I would just post this trips anyone else up 😄

**Comment by TMBL-DEV** [comment]
My dev server is running but when i do `yarn run nitro task list`

then it results in 
` Missing info file: /home/timmy/projects/private/rewrite/nuxt-project/.nitro/nitro.json (is dev server running?) `

what bothers mee is that it's looking in .nitro folder whilst nuxt generates a .nuxt/nitro.json.

does anyone have a fix?

**Comment by ProgramRipper** [comment]
> > @robin-dongbin You mean(意味着) this [nitro.unjs.io/guide/tasks#scheduled-tasks](https://nitro.unjs.io/guide/tasks#scheduled-tasks) or something different? (like a pattern(模式) for <1 minute(分钟) schedule?)
> 
> Yes, that's what I want. Is it possible? I know that cron do not support that, but Laravel has support this last year.

Yes, it is supported. You can use cron expression like this `*/15 * * * * *` (six `*`, different from five `*` of classic cron), the first `*` means seconds.

**Comment by robin-dongbin** [comment]
> > > @robin-dongbin You mean(意味着) this [nitro.unjs.io/guide/tasks#scheduled-tasks](https://nitro.unjs.io/guide/tasks#scheduled-tasks) or something different? (like a pattern(模式) for <1 minute(分钟) schedule?)
> > 
> > 
> > Yes, that's what I want. Is it possible? I know that cron do not support that, but Laravel has support this last year.
> 
> @robin-dongbin Yes, it is supported. You can use cron expression like `*/15 * * * * *` (six `*`, different from five `*` of classic cron), the first `*` means seconds.

That's great, thank you!

**Comment by flapili** [comment]
Hi, do you think in the future this feature could allow us to define scheduled function on gcp as example ?

**Comment by pi0** [comment]
@flapili Sure 👍🏼  if you can perhaps share an example project and some links to the docs i can help to do it faster :)

**Comment by flapili** [comment]
> @flapili Sure 👍🏼 if you can perhaps share an example project and some links to the docs i can help to do it faster :) (we might need a new preset that auto configured for gcp)


the usecase we encounter at work is about run every day a routine to check is point of sales are closed, if not send an email, after 6 day force close the POS (constraint from France law)

here an example from google directly : https://github.com/firebase/functions-samples/blob/a8b1ab89c503419fab90b0a44b61adee8ec9923b/Node/delete-unused-accounts-cron/functions/index.js#L37

**Comment by pi0** [comment]
I see so we can probably integrate it [firebase preset](https://nitro.unjs.io/deploy/providers/firebase).

**Comment by Eckhardt-D** [comment]
Really cool feature! Works like a charm in VPS environments for scheduledTasks. I have 2 questions:

If for example, the nitro app is running in a cluster (e.g. PM2). I am assuming the scheduled tasks will trigger in each instance on the cron schedule? What would you suggest here?

Will scheduled tasks hook into platform API's for jobs, e.g. Firebase has scheduled functions and Vercel has Cron Jobs? Both which accept the cron expressions.

**Comment by Niki2k1** [comment]
@Eckhardt-D pi0 wrote about cluster mode in this comment: https://github.com/unjs/nitro/issues/1974#issuecomment-1967387326

**Comment by ivanjeremic** [comment]
I don't see in the experimental docs a way to programmatically add/remove tasks, for example, add a task programmatically that executes at a specific time, is that doable?

**Comment by pi0** [comment]
By design, no it is not doable to define or modify the tasks in runtime. The reason is that we need deterministic behavior during build to integrate with platform primitives.


@ivanjeremic Can you explain more of the usecases that might require this dynamic behavior?

**Comment by manniL** [comment]
@ivanjeremic couldn't you run a job every X minutes that checks if some posts are queued up, then processing them if so? 🤔

**Comment by ivanjeremic** [comment]
@manniL I wanted to say I'm new to tasks/cron and maybe the way I think about this problem is wrong, I think you are right and the way you described it is a simpler way of doing what I want.

**Comment by pi0** [comment]
Thanks for explaining your idea @ivanjeremic. I think in the future something like this would unblock this and i have added two tasks to tracker.

```js
runTask('useTask', {
  scheduled: 5 * 60, /* 5 minutes */
  multiple: true,
  payload: {
    userId: 123
  }
})
```

(above snippet is just an example to show the idea)

**Comment by ivanjeremic** [comment]
> Thanks for explaining your idea @ivanjeremic. I think in the future something like this would unblock this and i have added two tasks to tracker.
> 
> 
> 
> ```js
> 
> runTask('useTask', {
> 
>   scheduled: 5 * 60, /* 5 minutes */
> 
>   multiple: true,
> 
>   payload: {
> 
>     userId: 123
> 
>   }
> 
> })
> 
> ```
> 
> 
> 
> (above snippet is just an example to show the idea)

Exactly how I imagined it👍🏻

**Comment by iBobik** [comment]
It would be handy to disable cron tasks in dev server by default, but keep ability to run it manually. It is annoying when I need to debug something else and see log full of unrelated periodic tasks.

**Comment by casualmatt** [comment]
It's very nice how the tasks show up in nuxt devtools; I tried to use `h3 `utility to send a stream back but didn't succeed because they need the `event`; this could be an excellent tool to keep track of the task activity.

**Comment by malgamves** [comment]
New to this so curious if it normal that when i setup a task in Nuxt to return something and use the cron scheduler in my Nuxt config, my tasks run but i don't get the return object i'm expecting. 

When i run the task programmatically, say through an api route in Nuxt, i get the return object.

**Comment by jsonleex** [comment]
Thanks for the awesome feature! 🫰

Is there any way to create a new task within a task, or can tasks be created during runtime?

**Comment by AwesomePeregrineFalcon** [comment]
Do you want to add the ability to run tasks from CLI? Like
```bash
node .output/server/index.mjs db:migrate
```
or
```bash
node .output/server/index.mjs task run db:migrate
```

I really liked Nitro but this feature is kinda deal-breaker. As I understand, there is no way right now to make such commands

**Comment by pi0** [comment]
@AwesomePeregrineFalcon it is in the plan yes to allow CLI interaction with production build. In the meantime you can expose a (protected) API endpoint.

**Comment by juretopolak** [comment]
> The only thing that i need to figure out is how to get cloudflare d1 working in the tasks context

@TMBL-DEV have you figured this out? 🙂

**Comment by AwesomePeregrineFalcon** [comment]
> it is in the plan yes to allow CLI interaction with production build. In the meantime you can expose a (protected) API endpoint.

@pi0, I figured out how to make CLI. I made the custom preset. Just used `Commander.js` library in `server/preset/entry.ts` and made the listening start in a subcommand, so I start the server with
```bash
node .output/server/index.mjs server listen
```
and run migrations with
```bash
node .output/server/index.mjs db migrate
```
But now I have to build the project to use CLI :disappointed:. Anyway, it works now... And I liked everything else in Nitro.js! Is there a reason why there is no possibility to run a custom preset in dev mode? Something like:
```bash
yarn dev server listen
```

**Comment by dalbodeule** [comment]
Wouldn't it be possible to bind specific tasks to the Task API? For example, it might be a good idea to apply cloudflare worker's email handler or queue handler.

While looking at related issues and PRs, I thought it was a good idea and suggested it.

**Comment by luke-z** [comment]
I just came across the Tasks API and was wondering if eventually it could replace a package like bullmq for simple jobs and a worker type of setup.

> Allow optionally allow running multiple instances of task

Seems to get close to this kind of feature.

Or am I somehow misunderstanding it :)

**Comment by kyng-cytro** [comment]
Hey, does the cron Synatx have support for time zones? I know Croner does.

**Comment by provok-me** [comment]
> If I use a nitro task in, let's say, Netlify will it leverage netlify background functions?

@pi0 
A Nitro task has only one running instance, but Netlify background functions can handle multiple instances at the same time (as I understand it in there docs).

Is there another way to use Netlify background functions with Nitro to handle multiple instances of a background function?

**Comment by StepanMynarik** [comment]
Using Nuxt 3.

Tried building an endpoint to get scheduled jobs using 'scheduledJobs' variable imported from "cron" package directly, see here: https://croner.56k.guru/usage/examples/#naming-jobs
My intention was to build a UI that shows the scheduled jobs, ideally even which ones are currently running with which payload etc.

The 'scheduledJobs' variable however returns an empty array always.

Is there a way to do this? To monitor scheduled and running jobs?

**Comment by luc122c** [comment]
Hi @pi0, I'm wondering how scheduled tasks are handled in a serverless context? For example, I am currently using the Firebase preset which uses a Firebase Function as the 'server'. Since the server is not long-lived, would scheduled tasks execute each time the service is called, or are they not included in unsupported presets?

**Comment by gbyesiltas** [comment]
hey, is it currently possible to create a nuxt application with tasks and get it to work with cloudflare triggers? (as it's deployed as a pages application I couldn't get it to work)

**Comment by carlos-duran** [comment]
Hello, I was trying to run `nitro task run hi` in a Nuxt 3 app but it looks like the `.nitro` folder is expected. We have `.nuxt` folder.

Btw, Thank you for a great API!

**Comment by TMBL-DEV** [comment]
@carlos-duran i had the same issue i have an open pull request at the moment that should fix it

https://github.com/unjs/nitro/pull/2635

**Comment by aligzl** [comment]
> hey, is it currently possible to create a nuxt application with tasks and get it to work with cloudflare triggers? (as it's deployed as a pages application I couldn't get it to work)
> 
> update: i got it to work after deploying the nuxt app with [the cloudflare_module preset here](https://nitro.unjs.io/deploy/providers/cloudflare#cloudflare-module-workers) which deploys it as a workers application

Can you please give an example how you did it?
Thanks

**Comment by gbyesiltas** [comment]
>Can you please give an example how you did it?
Thanks

@aligzl hey, I pretty much exactly followed [the docs for the cloudflare_preset module](https://nitro.unjs.io/deploy/providers/cloudflare#cloudflare-module-workers) which I believe builds/deploys the app as a workers application and not as a pages application. Nitro tasks started working for me after that

**Comment by aligzl** [comment]
hmm, i am using nuxt hub, i guess i will mess it up if i touch it :)

**Comment by zacwebb** [comment]
> hmm, i am using nuxt hub, i guess i will mess it up if i touch it :)

I'm in the same boat with NuxtHub. With the new Cloudflare worker updates, support for this will be coming soon to NuxtHub I believe (cc @atinux 👀)

**Comment by ymansurozer** [comment]
> The only thing that i need to figure out is how to get cloudflare d1 working in the tasks context

The `context` inside the `TaskEvent` seems to be the same as a usual `context` in an API route for Cloudflare Workers deployment. So you can access the Cloudflare env in that, too. @TMBL-DEV @juretopolak

**Comment by gsxdsm** [comment]
Love to see this can't wait until it's fully ready

**Comment by shoooe** [comment]
I tried this 3 weeks ago and now I can't make them work anymore for some reason following the same guide I had specified here: https://github.com/solidjs/solid-docs/issues/964

I also tried in Tanstack Start which is also based on Nitro and it also doesn't work there.

Did something change somehow in the last 3 weeks 😕? Seems like the last release was more than 3 weeks ago so I'm assuming no.

Anybody else has the same problem?

Or if you made it work could you post your `app.config.ts` and `tasks` folder?

**Comment by pi0** [comment]
@shoooe both tanstack-start and solid are based on a wrapper "Vinxi" which sadly abstracts away full features of Nitro (consider even dev server of nitro is not used). Currently Nuxt, AnalogJS or pure Nitro projects can use all features. Hopefully this changes in the near future for them.

**Comment by kstraszewski** [comment]
Im using Nitro Tasks on Nuxt3, and i have an issue when i deploy them on Vercel. Everything works great locally, but when i deploy app on vercel the Task mysteriously dies and i do not have any information why. When i've added console.logs to the function i see that at some point they stops. Any idea how to debug it, or why it might happened?

**Comment by luc122c** [comment]
> on vercel the Task mysteriously dies and i do not have any information why. 

I would imagine it's [timing out](https://vercel.com/guides/what-can-i-do-about-vercel-serverless-functions-timing-out). Vercel is not really the best place for long-lived tasks.

**Comment by kstraszewski** [comment]
> > on vercel the Task mysteriously dies and i do not have any information why.
> 
> I would imagine it's [timing out](https://vercel.com/guides/what-can-i-do-about-vercel-serverless-functions-timing-out). Vercel is not really the best place for long-lived tasks.

I have this in my nuxt.config

```json
...
    vercel: {
      functions: {
        maxDuration: 120,
      },
    },    
 ```
 Task should take about 30-45sec.
 I do not see anything in the vercel logs.

**Comment by pi0** [comment]
Added vercel support to tracker 🤞🏼 

In the meantime, you can instruct nitro config and create a custom handler for vercel:


```js
export default defineNitroConfig({
  vercel: {
    config: {
      crons: [
        {
          // TODO: create server/api/_cron.ts event handler
          path: "/api/_cron", 
          schedule: "0 0 * * *",
        }
      ],
    },
  },
});
```

**Comment by gritworks** [comment]
thanks pi0.

Also seeing support for this way of defining the cronjob in the nuxt.config file.
I guess this will work also the same way:

nitro: {
      vercel:{
        config:{
            "crons": [{
            "path": "/tasks/mytask",
            "schedule": "0 0 2 * *"
          }]
        }
      } 
  }

**Comment by medz** [comment]
@pi0 I have a question. I use Nitro development server and deploy it with docker. There may be multiple pods. Do we have any way to ensure that the task will only run once?

For example, Redis shared lock, etc.

**Comment by ooxoxx** [comment]
@pi0 hi, thanks for the hard work on this feature. I really need that for a data scrawling scenario. Is there any workarounds about long-running tasks on production environments like netlify right now?
Here is my major requirements:
1. single netlify function instance for caching/state-control. (Tasks result in cache/state changing in application level.)
2. get rid of 10 seconds limit of api calling from netlify function. (maybe take advantage of their background function feature?)
I think it's literally not a very simple quest. Should we make the entire server.mjs function a "background" one?

**Comment by tomhaerter** [comment]
I'm curious if there's any plan for a queue system like [Rails Solid Queue](https://github.com/rails/solid_queue/) or [Laravel Queues](https://laravel.com/docs/master/queues). (I saw the previous [comments](https://github.com/nitrojs/nitro/issues/1974#issuecomment-1978213515) about Laravel Queues)

Things like these would be great:

- Queue Worker(s) _(which can run independently of the main Nitro server, for horizontal scaling see [Rails Solid Queue](https://github.com/rails/solid_queue/#:~:text=For%20small%20projects%2C%20you%20can%20run%20Solid%20Queue%20on%20the%20same%20machine%20as%20your%20webserver.%20When%20you%27re%20ready%20to%20scale%2C%20Solid%20Queue%20supports%20horizontal%20scaling%20out%2Dof%2Dthe%2Dbox.))_
- Storage Adapters (SQL, Redis, RabbitMQ, Amazon SQS, etc.)
- Retrying of tasks

**Comment by Larsmyrup** [comment]
Hi,

Is this supposed to work on Netlify out of the gate?

I recently deployed a Nuxt project to Netlify, and pretty much everything, but the scheduled tasks, seems to work as expected. The scheduled tasks are running and working as expected locally.

Due to the feature currently being experimental, i just wanted to confirm that it's suppose to work as is. Has anyone had success making scheduled tasks work on Netlify? 

Thanks!

**Comment by jamesarosen** [comment]
Two task providers that I don't see mentioned here are [DBOS](https://www.dbos.dev/) and [Trigger.dev](https://trigger.dev/). Both are great ways of implementing _durable execution_, which is an awesome pattern for background work. DBOS requires Postgres, but otherwise has no infrastructure dependencies. Trigger requires at least one additional Docker container to operate. Both are great if you're hosting your app on a platform like Fly.io or AWS ECS where you have long-lived containers.

If there's interest, I'd be happy to help with a DBOS adapter.

**Comment by schplitt** [comment]
I have created a PR to support scheduled one time tasks [here](https://github.com/nitrojs/nitro/pull/3237)

**Comment by medz** [comment]
It would be even better if it could support cloudflare~

**Comment by alexcroox** [comment]
How do I get access to `useRuntimeConfig` in Nitro tasks?

I have a cron job that needs to connect to a database and my db function uses:

```
const { dbHost, dbUser, dbPassword } = useRuntimeConfig(event)
```

However Nitro tasks don't expose an event to pass here, and using `useRuntimeConfig` directly in the task `run()` function doesn't work either.

**Comment by luc122c** [comment]
> How do I get access to `useRuntimeConfig` in Nitro tasks?

I have been using `useEvent`, but you need to enable [async context](https://nitro.build/guide/utils#async-context-experimental) to use this.

```ts
async run({ payload, context }) {
	const event = useEvent();
	const config = useRuntimeConfig(event)
}
```

I am not sure if it will always be available (e.g. if triggered by cron), but it seems to have been working for me for tasks run with `runTask`.

**Comment by grokpot** [comment]
## Nuxt + Vercel: Working Solution with Nitro Tasks

**What did not work for me:**
- Direct Vercel cron to `/tasks/myTask` → 404 errors in production

**What worked for me:**
Custom API wrapper that calls `runTask()`:

```typescript
// nuxt.config.ts
vercel: {
  config: {
    crons: [{ schedule: '0 11 * * *', path: '/api/vercelCron/myTask' }]
  }
},
supabase: { // in case you're using @nuxtjs/supabase, important you exclude the api from auth
  redirectOptions: {
    exclude: ['/api/vercelCron/**']
  }
}

// server/api/vercelCron/myTask.ts  
/* delete this file when nitro team fixes vercel task integration */
export default defineEventHandler(async (event) => {
  const config = useRuntimeConfig();
  
  // Vercel CRON_SECRET security check
  if (getHeader(event, 'authorization') !== `Bearer ${config.cronSecret}`) {
    throw createError({ statusCode: 401 });
  }

  // Call Nitro task
  const { result } = await runTask('myTask');
  return { success: true, result };
});

// server/tasks/myTask.ts (unchanged)
export default defineTask({ /* your task logic */ });
```

**Benefits:**
- ✅ Uses Nitro tasks (gets concurrency control, dev tools, etc.)
- ✅ Works on Vercel production  
- ✅ Follows Vercel security practices
- ✅ Can easily switch back when Vercel supports tasks directly

**Environment:** Nuxt 3, Vercel, `@nuxtjs/supabase`

**Comment by jd-solanki** [comment]
Is anyone facing this error in nuxt v4?

```
ERROR  Missing info file: /Users/___/.nitro/nitro.json (is dev server running?)
```

I have `nitro.json` in `.nuxt/nitro.json`. Maybe nitro is looking at the wrong file.

**Comment by acidjazz** [comment]
should the task CLI work in nuxt?
```
❯ npx nitro task list

 WARN  Please add compatibilityDate: '2025-07-29' to the config file. Using 2024-04-03 as fallback.                  nitro 4:19:20 PM
       More info: https://nitro.build/deploy#compatibility-date


 ERROR  Missing info file: /Users/k/vc/site/.nitro/nitro.json (is dev server running?)                                     4:19:20 PM

    at _getTasksContext (node_modules/.pnpm/nitropack@2.12.3_@libsql+client@0.14.0_@netlify+blobs@9.1.2_better-sqlite3@11.9.1_drizz_8d96dadb3049623190a0290c9f38bdaf/node_modules/nitropack/dist/core/index.mjs:1034:11)
    at listTasks (node_modules/.pnpm/nitropack@2.12.3_@libsql+client@0.14.0_@netlify+blobs@9.1.2_better-sqlite3@11.9.1_drizz_8d96dadb3049623190a0290c9f38bdaf/node_modules/nitropack/dist/core/index.mjs:989:21)
    at Object.run (node_modules/.pnpm/nitropack@2.12.3_@libsql+client@0.14.0_@netlify+blobs@9.1.2_better-sqlite3@11.9.1_drizz_8d96dadb3049623190a0290c9f38bdaf/node_modules/nitropack/dist/cli/list.mjs:20:25)
    at async runCommand (node_modules/.pnpm/citty@0.1.6/node_modules/citty/dist/index.mjs:316:16)
    at async runCommand (node_modules/.pnpm/citty@0.1.6/node_modules/citty/dist/index.mjs:307:11)
    at async runCommand (node_modules/.pnpm/citty@0.1.6/node_modules/citty/dist/index.mjs:307:11)
    at async runMain (node_modules/.pnpm/citty@0.1.6/node_modules/citty/dist/index.mjs:445:7)



 ERROR  Missing info file: /Users/k/vc/site/.nitro/nitro.json (is dev server running?)                                     4:19:20 PM

~/vc/site xdb-import* ❯
```

**Comment by medz** [comment]
In preset: "cloudflare_module", can task events inject bindings and vars into the context during local development? This difference prevents local debugging of tasks in programs deployed on workers.

**Comment by Mokkapps** [comment]
Love that feature, I currently use unmaintained Nuxt modules to schedule such tasks and would love to switch to tasks instead.

I'm missing the functionality to add timezones, it is supported by Croner but I see no option to pass it to the task?

**Comment by medz** [comment]
> In preset: "cloudflare_module", can task events inject bindings and vars into the context during local development? This difference prevents local debugging of tasks in programs deployed on workers.

```typescript
export function useCloudflareEnv(event?: H3Event | TaskEvent): CloudflareEnv {
  try {
    const { context } = event ?? useEvent();
    return (context as any).cloudflare.env as CloudflareEnv;
  } catch {
    // https://github.com/nitrojs/nitro/blob/v2/src/presets/cloudflare/runtime/_module-handler.ts#L37
    return (globalThis as any).__env__;
  }
}
```

Temporarily solved the problem

**Comment by aburii** [comment]
> should the task CLI work in nuxt?
> 
> ```
> ❯ npx nitro task list
> 
>  WARN  Please add compatibilityDate: '2025-07-29' to the config file. Using 2024-04-03 as fallback.                  nitro 4:19:20 PM
>        More info: https://nitro.build/deploy#compatibility-date
> 
> 
>  ERROR  Missing info file: /Users/k/vc/site/.nitro/nitro.json (is dev server running?)                                     4:19:20 PM
> 
>     at _getTasksContext (node_modules/.pnpm/nitropack@2.12.3_@libsql+client@0.14.0_@netlify+blobs@9.1.2_better-sqlite3@11.9.1_drizz_8d96dadb3049623190a0290c9f38bdaf/node_modules/nitropack/dist/core/index.mjs:1034:11)
>     at listTasks (node_modules/.pnpm/nitropack@2.12.3_@libsql+client@0.14.0_@netlify+blobs@9.1.2_better-sqlite3@11.9.1_drizz_8d96dadb3049623190a0290c9f38bdaf/node_modules/nitropack/dist/core/index.mjs:989:21)
>     at Object.run (node_modules/.pnpm/nitropack@2.12.3_@libsql+client@0.14.0_@netlify+blobs@9.1.2_better-sqlite3@11.9.1_drizz_8d96dadb3049623190a0290c9f38bdaf/node_modules/nitropack/dist/cli/list.mjs:20:25)
>     at async runCommand (node_modules/.pnpm/citty@0.1.6/node_modules/citty/dist/index.mjs:316:16)
>     at async runCommand (node_modules/.pnpm/citty@0.1.6/node_modules/citty/dist/index.mjs:307:11)
>     at async runCommand (node_modules/.pnpm/citty@0.1.6/node_modules/citty/dist/index.mjs:307:11)
>     at async runMain (node_modules/.pnpm/citty@0.1.6/node_modules/citty/dist/index.mjs:445:7)
> 
> 
> 
>  ERROR  Missing info file: /Users/k/vc/site/.nitro/nitro.json (is dev server running?)                                     4:19:20 PM
> 
> ~/vc/site xdb-import* ❯
> ```

It is, I guess.

For now you can create a `nitro.config.ts` file as a workaround and specify a `buildDir` to `.nuxt`.

It will create a warning when launching the Nuxt dev server but still works.

**Comment by alanscodelog** [comment]
`useEvent` does not work for me when the task is run as a cron job (asyncContext is on), I get the context not available error. It does work when triggering it manually via the web or cli (I assume because in that case there *is* an event being sent). `useRuntimeConfig` always works fine regardless.

Really what I want the context for is the database connection. I can get it manually, but kind of annoying.

Also small side note, I could not find how to change the host/port of the cli and I'm usually running with locally signed https certificates so it doesn't work. Also even though `asyncContext` is on, I get a type error on the type of the run function.

**Comment by MiniDigger** [comment]
a hook for tasks would be pretty useful, so that you can add stuff to the task context via a plugin, if thats something people think is useful I could try PRing it

**Comment by tomhaerter** [comment]
would it be possible to use https://useworkflow.dev/ to have the nitro tasks api work durably?

**Comment by ijsnow** [comment]
Have I missed some guidance for getting tasks to use with vite somewhere? It appears to be possible because the server logged an error about an unregistered task as I had the srcDir configured wrong. Here is what I have:

```typescript
// vite.config.ts

import tailwindcss from "@tailwindcss/vite"
import { tanstackStart } from "@tanstack/react-start/plugin/vite"
import viteReact from "@vitejs/plugin-react"
import { defineConfig } from "vite"
import viteTsConfigPaths from "vite-tsconfig-paths"
import { nitro } from "nitro/vite"

const config = defineConfig({
  plugins: [
    viteTsConfigPaths({
      projects: ["./tsconfig.json"],
    }),
    tailwindcss(),
    tanstackStart(),
    viteReact(),
    nitro({
      config: {
        preset: "node-server",
        srcDir: "src",
        experimental: {
          tasks: true,
        },
        scheduledTasks: {
          "*/10 * * * *": ["hello"],
        },
      },
    }),
  ],
})

export default config
```

```typescript
// src/tasks/hello.ts

export default defineTask({
  meta: {
    name: "hello",
    description: "Say hello.",
  },
  run({ payload, context }) {
    console.log("Hello")

    return { result: "Success" }
  },
})
```

Running `pnpm vite dev` starts my server and serves the site but doesn't log `Hello` every 10 seconds as I'd expect

**Comment by milan-codes** [comment]
> Running `pnpm vite dev` starts my server and serves the site but doesn't log `Hello` every 10 seconds as I'd expect

@ijsnow Cron only allows for a minimum of one minute. What you specified for the scheduled task is every 10th minute. That might have been the problem, but I'm not sure as I'm struggling to make this feature of Nitro work with TanStack Start, too.

**Comment by Cry0nicS** [comment]
> **What worked for me:** Custom API wrapper that calls `runTask()`:
> 
> // nuxt.config.ts
> vercel: {
>   config: {
>     crons: [{ schedule: '0 11 * * *', path: '/api/vercelCron/myTask' }]
>   }
> },
> supabase: { // in case you're using @nuxtjs/supabase, important you exclude the api from auth
>   redirectOptions: {
>     exclude: ['/api/vercelCron/**']
>   }
> }


Hey  @grokpot 
I tried your solution, but the `vercel` property in `nuxt.config.ts` on `nitro`? Like 
```ts
// nuxt.config.ts
    nitro: {
        experimental: {
            asyncContext: true,
            tasks: true
        },
        vercel: {
            config: {
                crons: [{path: "/api/vercel/tasks", schedule: "0 20 * * *"}]
            }
        }
    },
```
Because otherwise I'm running into this error, because `vercel` doesn't seem to exist.
```
TS2353: Object literal may only specify known properties, and 'vercel' does not exist in type 'InputConfig<NuxtConfig, ConfigLayerMeta>'.
```
[Docs](https://nuxt.com/deploy/vercel#custom-build-output-configuration) also hint at the same:
>You can provide additional [build output configuration](https://vercel.com/docs/build-output-api/v3) using nitro.vercel.config key inside nuxt.config.ts.

**Comment by grokpot** [comment]
@Cry0nicS, yes, here's a copy-paste of my code:
```
nitro: {
    experimental: {
      tasks: true,
    },
    // https://github.com/nitrojs/nitro/issues/1974
    // When fixed, remove the 'vercel key and move to 'scheduledTasks'
    vercel: {
      config: {
        crons: [
          {
            schedule: '0 11 * * *',
            path: '/api/vercelCron/generateDailyTextChallenge',
          },
          {
            schedule: '5 11 * * *', // 5 minutes later to avoid conflicts
            path: '/api/vercelCron/generateDailyImageChallenge',
          },
        ],
      },
    },
  },
```

package.json:
"nuxt": "^4.1.2"

**Comment by ilkergzlkkr** [comment]
> Have I missed some guidance for getting tasks to use with vite somewhere? It appears to be possible because the server logged an error about an unregistered task as I had the srcDir configured wrong. Here is what I have:

Adding `scanDirs` instead of `srcDir` to vite config fixed my issue.

```ts
nitro: {
    scanDirs: ["src"],
    ...
}
```

**Comment by aburii** [comment]
Would be awesome to have an option to set the scheduled task to a specific timezone!

**Comment by hareland** [comment]
> Would be awesome to have an option to set the scheduled task to a specific timezone!

The timezone of the cron expression is depending on the system timezone afaik.

**Comment by MickL** [comment]
@pi0 I do have some questions regarding tasks:

1. Are the [v3 docs](https://v3.nitro.build/docs/tasks) still valid for v3? Including the need of setting `experimental.tasks` ?
2. If the task name is generated by file name, e.g. `tasks/db/migrate.ts` -> `db:migrate`, why does `defineTask()` require to set `meta.name`? (possibly update docs with answer)
3. How is the naming resolved? E.g. `tasks/cleanup-sessions.ts` -> `cleanup-session` or `cleanup:session`? (possibly update docs with answer)
4. Why can tasks return values? I expected them to run only internally by a scheduler. (possibly update docs with answer)
5. Personally I tried to add a task to my v3 project but it doesnt seem to work. They show up on `_nitro/tasks` but they dont run (just testing in dev):
```json
{
    "tasks": {
        "cleanup-sessions": {
            "description": "Clean sessions that were not marked completed/expired by external webhook"
        }
    },
    "scheduledTasks": [
        {
            "cron": "*  *  *  *  *",
            "tasks": [
                "cleanup-sessions"
            ]
        }
    ]
}
```

**Comment by drewbaker** [comment]
> Thanks for explaining your idea [@ivanjeremic](https://github.com/ivanjeremic). I think in the future something like this would unblock this and i have added two tasks to tracker.
> 
> runTask('useTask', {
>   scheduled: 5 * 60, /* 5 minutes */
>   multiple: true,
>   payload: {
>     userId: 123
>   }
> })

I would love this, but would want `scheduled: "now"` to have the task run right away. And I'd really need the ability to run multiple's of the same task (with different payloads). My use case is something similar to "import this data that might take 10 minutes".

**Comment by hareland** [comment]
Hi @MickL,

Don't have the answer to all questions, but:
4. Tasks can return a value since [you can also call it programmatically](https://v3.nitro.build/docs/tasks#programmatically-run-tasks), a task is also required to return a result as `return {result: ...}`

**Comment by hareland** [comment]
Hi @drewbaker, it seems you are looking for a more "queue" style handling of tasks, as you cannot dispatch a task to run in the background afaik. 
I got a **feeling** that [vercel workflow](https://github.com/vercel/workflow) will be added to handle such usecases in the future, or you can add it yourself.

**Comment by drewbaker** [comment]
@hareland Yeah, I've actually got Sidequest.js [working with Nitro](https://github.com/nitrojs/nitro/discussions/4086) (sort of) so I have the queue part solved, I just need a way to fire off Nitro functions from within the worker jobs, and I figured Tasks might have been the best way to do that.

**Comment by jd-solanki** [comment]
Hi @pi0 

Noticed, my nitro tasks from layer isn't merged/picked when I have root nuxt config having `nitro.cloudflare.wrangler`.

**Comment by BaptisteCrouzet** [comment]
Is this feature planned to be stable in V3 ? Or at least no more experimental status for the currently implemented part ?

**Comment by chenyuncai** [comment]
how about set schedule in task meta?

**Comment by hareland** [comment]
> how about set schedule in task meta? 

Personally I think it's not the tasks concern when it runs - you'll get config drift at some point..
